nvflare/app_common/streamers/object_retriever.py

```diff
@@ -0,0 +1,284 @@
+
+        Returns: tuple of (ReturnCode, Validation Data)
+        This method should do as much as possible so that the do_stream method won't be called if any error
+        is detected (the do_stream method is called in a separate thread).
+        The validation data produced by this method will be passed to the do_stream method.
+
+        """
+        pass
+
+    @abstractmethod
+    def do_stream(
+        self,
+        target: str,
+        request: Shareable,
+        fl_ctx: FLContext,
+        stream_ctx: StreamContext,
+        validation_data: Any,
+    ) -> Any:
+        """Object sending side. Called to stream data to the requesting side.
+
+        Args:
+            target: the requesting site to stream to
+            request: the object retrieval request
+            fl_ctx: a FLContext object
+            stream_ctx: stream context data
+            validation_data: the validation data produced by the validate_request method.
+
+        Returns: Any object
+
+        """
+        pass
+
+    @abstractmethod
+    def get_result(self, stream_ctx: StreamContext) -> (str, Any):
+        """Object requesting side, which is also the stream receiving side.
+        Called to get the result of the streaming.
+
+        Args:
+            stream_ctx: StreamContext object
+
+        Returns: tuple of (ReturnCode, Result Object)
+
+        """
+        pass
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.START_RUN:
+            engine = fl_ctx.get_engine()
+            self.register_stream_processing(
+                fl_ctx=fl_ctx,
+                channel=self.stream_channel,
+                topic=self.topic,
+                stream_done_cb=self._handle_stream_done,
+            )
+            engine.register_aux_message_handler(topic=self.topic, message_handle_func=self._handle_request)
+
+    def retrieve(self, from_site: str, fl_ctx: FLContext, timeout: float, **obj_attrs) -> (str, Any):
+        """Retrieve an object from a specified site.
+
+        Args:
+            from_site: the site to retrieve the object from
+            fl_ctx: a FLContext object
+            timeout: max number of seconds to wait for the data
+            **obj_attrs: attributes of the object to be retrieved
+
+        Returns: tuple of (ReturnCode, Retrieved Data)
+
+        """
+        engine = fl_ctx.get_engine()
+        waiter = _Waiter()
+        tx_id = str(uuid.uuid4())
+        self.tx_table[tx_id] = waiter
+        self.log_debug(fl_ctx, f"set waiter for Rtr {tx_id}")
+
+        request = Shareable({RETRIEVER_TX_ID: tx_id})
+        if obj_attrs:
+            request.update(obj_attrs)
+
+        # ask the site to start streaming
+        reply = engine.send_aux_request(
+            targets=[from_site], request=request, topic=self.topic, fl_ctx=fl_ctx, timeout=timeout
+        )
+        # the reply is a dict keyed with site names!
+        reply = reply.get(from_site)
```

Comment:
It's confusing to redefine the vars here. Should use 2 different vars here. Maybe replies and reply.

Reply:
Could be.